readme.txt
PERFORCE DEFECT TRACKING INTEGRATION README FOR RELEASE 2.4.5

Richard Brooksby, Ravenbrook Limited

$Date: 2009/05/07 $


CONTENTS

  1. Introduction
  2. Installation
    2.1. Upgrading from an old release of the P4DTI
    2.2. Installing the Bugzilla integration on Linux
    2.3. Installing the Bugzilla integration on Solaris
    2.4. Installing the Bugzilla integration on Windows 2000
    2.5. Installing the Integration Kit
  A. References
  B. Document history
  C. Copyright and license

1. INTRODUCTION

This is release 2.4.5 of the Perforce Defect Tracking Integration
(P4DTI).

The P4DTI connects your defect tracking system to Perforce, so that you
don't have to switch between them and enter duplicate information about
your work.  It also links changes made in Perforce with defect tracker
issues, making it easy to find out why a change was made, find the work
that was done to resolve an issue, or generate reports relating issues
to files or codelines.

For supported configurations, contact information, and what's new in
this release, see the release notes (release-notes.txt).

If you are planning to modify, adapt, or extend the P4DTI, please
download and unpack the P4DTI Integration Kit.  See section 2.5.

The readership of this document is anyone who wants to download and use
the Perforce Defect Tracking Integration.

This document is not confidential.


2. INSTALLATION


2.1. Upgrading from an old release of the P4DTI

If you're running an old release of the P4DTI, then you must follow
these steps when upgrading to this release:

  1. Stop the replicator.

  2. Make a copy of your configuration file config.py in a safe place.

  3. (Bugzilla only.)  Remove the old Bugzilla patch.  Go to the Bugzilla
installation directory and run the command:

     patch -R < /opt/p4dti/bugzilla-VERSION-patch

where VERSION is the version of Bugzilla (for example, 3.0).  (If
you didn't install the Bugzilla integration as an RPM, then the patch
file will be wherever you installed it, not necessarily in
/opt/p4dti.)

  4. If your old release was installed as an RPM, run the following
command as root to uninstall it:

     rpm -e p4dti

  5. Install the P4DTI as normal (see sections 2.2 to 2.4 below).

  6. Copy your old configuration file as config.py in the new P4DTI
installation directory.


2.2. Installing the Bugzilla integration on Linux

The integration is distributed as an RPM called
"p4dti-2.4.5-1.i386.rpm".  Transfer this to the Bugzilla server machine,
then run the following command as root:

  rpm -i p4dti-2.4.5-1.i386.rpm

This will install the P4DTI files into /opt/p4dti and a startup script
in the /etc/rc.d/init.d directory.

Then consult the Administrator's Guide (installed as
"/usr/doc/p4dti-2.4.5/ag/index.html") for further instructions.

If you prefer not to use RPMs, you can follow the procedure in section
2.3.


2.3. Installing the Bugzilla integration on Solaris

The integration is distributed as a gzipped tar file called
"p4dti-bugzilla-2.4.5.tar.gz".  Unpack this tarball on the Bugzilla
server machine, using the command

  gunzip -c p4dti-bugzilla-2.4.5.tar.gz | tar xvf -

Then consult the Administrator's Guide (installed as
"p4dti-bugzilla-2.4.5/ag/index.html") for further instructions.


2.4. Installing the Bugzilla integration on Windows 2000 or Windows NT

On the Bugzilla Windows machine, run the "p4dti-bugzilla-2.4.5.exe"
installer program that came with this document.  This will unpack the
integration.  We recommend installing the integration in "C:\Program
Files\P4DTI-2.4.5\" but you can ask the installer to put it somewhere
else.

Then consult the Administrator's Guide (installed by default in
"C:\Program Files\P4DTI-2.4.5\ag\index.html") for further instructions.


2.5. Installing the integration kit

The P4DTI integration kit is a complete set of P4DTI sources and
documentation to enable third parties to develop new integrations.
You should only install it if you are planning to modify, adapt, or
extend the P4DTI.

For Windows, the integration kit is distributed as the ZIP archive
"p4dti-kit-2.4.5.zip".  Unpack it using WinZip.

For Unix, the integration kit is distributed as the tarball
"p4dti-kit-2.4.5.tar.gz".  Unpack it using the command "gunzip -c
p4dti-kit-2.4.5.tar.gz | tar xvf -".

Then consult the Integration Kit readme file
(p4dti-kit-2.4.5/kit-readme.txt) for further instructions.


A. REFERENCES

None.


B. DOCUMENT HISTORY

2007-08-16 NB Updated for release 2.4.1

2007-09-25 NB Updated for release 2.4.2

2009-03-03 NB Updated for release 2.4.3

2009-04-14 NB Updated for release 2.4.4

2009-05-07 NB Updated for release 2.4.5

C. COPYRIGHT AND LICENSE

This document is copyright (C) 2001 Perforce Software, Inc.  All rights
reserved.

Redistribution and use of this document in any form, with or without
modification, is permitted provided that redistributions of this
document retain the above copyright notice, this condition and the
following disclaimer.

THIS DOCUMENT IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
HOLDERS AND CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
DOCUMENT, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

$Id: //info.ravenbrook.com/project/p4dti/version/2.4/readme.txt#7 $


release-notes.txt
PERFORCE DEFECT TRACKING INTEGRATION RELEASE NOTES FOR RELEASE 2.4.5

Nick Barnes, Ravenbrook Limited

$Date: 2009/05/07 $


CONTENTS

  1. Introduction
  2. Supported configurations
  3. Getting support
  4. Project contacts
  5. What's fixed
     5.1.   What's fixed in release 2.4.5
     5.2.   What was fixed in release 2.4.4
     5.3.   What was fixed in release 2.4.3
     5.4.   What was fixed in release 2.4.2
     5.5.   What was fixed in release 2.4.1
     5.6.   What was fixed in release 2.4.0
     5.7.   What was fixed in release 2.3.5
     5.8.   What was fixed in release 2.3.4
     5.9.   What was fixed in release 2.3.3
     5.10   What was fixed in release 2.3.2
     5.11.  What was fixed in release 2.3.1
     5.12.  What was fixed in release 2.3.0
     5.13.  What was fixed in release 2.2.2
     5.14.  What was fixed in release 2.2.1
     5.15.  What was fixed in release 2.2.0
     5.16.  What was fixed in release 2.1.2
     5.17.  What was fixed in release 2.1.1
     5.18.  What was fixed in release 2.1.0
     5.19.  What was fixed in releases up to 2.0.0
  A. References
  B. Document history
  C. Copyright and license


1. INTRODUCTION

These are the release notes for release 2.4.5 of the Perforce Defect
Tracking Integration (P4DTI).

The P4DTI connects your defect tracking system to Perforce, so that you
don't have to switch between them and enter duplicate information about
your work.  It also links changes made in Perforce with defect tracker
issues, making it easy to find out why a change was made, find the work
that was done to resolve an issue, or generate reports relating issues
to files or codelines.

For instructions on installing the P4DTI, see the product readme
(readme.txt).

For up-to-date information about releases of the P4DTI, see the product
information page <http://www.perforce.com/perforce/products/p4dti.html>.
From there you will find links to the latest releases, including
reports of defects found.

If you want to adapt or extend the P4DTI, please go to the product
information page <http://www.perforce.com/perforce/products/p4dti.html>
and download the Integration Kit.  It contains full source code and
documentation to help you.

The readership of this document is anyone who wants to download and use
the Perforce Defect Tracking Integration.

This document is not confidential.


2. SUPPORTED CONFIGURATIONS

Perforce versions:

  - between 2000.2 and 2008.2 inclusive, on any platform;

The Bugzilla for Unix integration supports:

  - Bugzilla 2.22.7 or 3.0.8 on Red Hat Linux or Solaris, storing the
    data in MySQL 4.0.26 or later;

  - Python 2.0, or later;

  - Running the P4DTI replicator on any Unix.

Use of Bugzilla 2.22.2 or 3.0 on Red Hat Linux or Solaris is also
supported, but is deprecated: support may be dropped in a future
version of the P4DTI.

Note that although the Bugzilla integration runs on a variety of Unix
operating systems, we've only tested it on FreeBSD 6.2 and Fedora Core
6.  Also note that the P4DTI will not be able to cause Bugzilla to
send email reporting bug changes unless it is running on the same
machine as Bugzilla itself.

The Bugzilla for Windows integration supports:

  - Bugzilla 2.22.7 or 3.0.8 on Windows NT4 or Windows 2000;

  - Python 2.0 or later with Win32 plug-ins;

  - Running the P4DTI replicator on Windows NT 4 or Windows 2000.

Use of Bugzilla 2.22.2 or 3.0 on Windows is also supported, but is
deprecated: support may be dropped in a future version of the P4DTI.

Integration with TeamShare's TeamTrack is no longer supported by
Perforce Software, or by this software.  An integration is available
from TeamShare.  Contact TeamShare support by writing to
<support@teamshare.com> or see the TeamShare web site at
<http://www.teamshare.com/> for further information.


3. GETTING SUPPORT

For problems relating to Perforce or the P4DTI in general, contact
Perforce Support by writing to <support@perforce.com> or see the
technical support page at
<http://www.perforce.com/perforce/support.html> for contact information.

Bugzilla is a community supported project.  See
<http://www.bugzilla.org/support>.


4. PROJECT CONTACTS

You may want to join the p4dti-discussion mailing list. The goals of the
list are:

  1. to provide feedback to the project on requirements, design,
implementation, etc.;

  2. to allow people to exchange information and experience with using
and adapting the project;

  3. to keep people informed about project progress.

To join, send a message with the word "subscribe" in the _body_ to
<p4dti-discussion-request@ravenbrook.com> or send the word "help" for
general information.

Please note that the mailing list will be archived and the archive may
be published.


5. WHAT'S FIXED

This section lists defects that have been fixed.


5.1. WHAT'S FIXED IN RELEASE 2.4.5


ESSENTIAL

job002109: non-ASCII on non-Unicode Perforce server breaks replicator

When non-ASCII characters are stored on a non-Unicode Perforce server
(e.g. by users entering them in a changelist description) the P4DTI
replicator doesn't know how to interpret them.  They are treated as raw
binary and then break when encoding as (e.g.) Latin-1 or ASCII.


5.2. WHAT WAS FIXED IN RELEASE 2.4.4


ESSENTIAL

job002084: P4DTI does no MySQL version detection

Some aspects of P4DTI Bugzilla operation depend on particular versions
of MySQL.  For instance, Unicode character set support was introduced
with MySQL 4.1.  Some more recent versions of MySQL are experimental or
beta releases and are not supported.  The P4DTI does not detect the
current version of MySQL, so cannot warn if the version is unsupported
or deprecated.


job002085: P4DTI doesn't check MySQL configured for Unicode

Bugzilla supports Unicode data, encoded as UTF-8 in MySQL, and since
Bugzilla 2.22 new installations are always UTF-8 encoded.  Since P4DTI
2.4.3 we replicate data as Unicode.  However, it is possible to run

logger.py
#              Perforce Defect Tracking Integration Project
#               <http://www.ravenbrook.com/project/p4dti/>
#
#                 LOGGER.PY -- PROGRAM LOGGING CLASSES
#
#             Gareth Rees, Ravenbrook Limited, 2000-10-16
#
#
# 1. INTRODUCTION
#
# This Python module implements classes for program logging -- recording
# information about the activity of a program.
#
# Logging is primarily intended to record what the integration does so
# that the administrator can:
#
#  1. undo it in an emergency [Requirements, 67];
#
#  2. debug their configuration [Requirements, 63];
#
#  3. remove the integration [Requirements, 64];
#
# and so that developers can:
#
#  4. debug modififcations of the system [Requirements, 25];
#
#  5. debug their extensions to new defect trackers [Requirements, 21].
#
# The intended readership of this document is project developers.
#
# This document is not confidential.

import catalog
import message
import os
import sys
import time
import types


# 2. ABSTRACT LOGGER CLASS
#
# "logger" is an abstract class for message logs, providing uniform
# message formatting and control of logging based on the priority of
# messages, but no actual logging.
#
# When constructing an instance of this class or a subclass, pass the
# minimum priority of message that should be written to the log (e.g.,
# message.DEBUG to see all debug-level messages, or message.ERROR to
# only see errors).

class logger:
    # Experienced a failure?
    has_failed = 0

    # Minimum priority of messages to log.
    priority = message.INFO

    # Maximum length of a log message.
    max_length = 10000

    def __init__(self, priority = message.INFO, max_length = 10000):
        assert isinstance(priority, types.IntType)
        self.priority = priority
        self.has_failed = 0
        self.max_length = max_length


    # 2.1. Convert a message to a string with truncation
    #
    # Truncate message to max_length to avoid very large log messages
    # causing the Windows event log or Linux system log to crash.

    def stringify_message(self, msg):
        assert isinstance(msg, message.message)
        return str(msg)[0:self.max_length]


    # 2.2. Format a message with date
    #
    # format_with_date(msg).  Format log message and return the message
    # as a string.  Messages are prefixed with the current date and time
    # in UTC and the message id.  See message.py for details of the
    # formatting of message ids.
    #
    # The reason for the existence of this method is that not all
    # logging methods record the date (in particular, file_logger
    # doesn't).

    def format_with_date(self, msg):
        assert isinstance(msg, message.message)
        date = time.strftime('%Y-%m-%d %H:%M:%S UTC',
                             time.gmtime(time.time()))
        return "%s  %s" % (date, self.stringify_message(msg))


    # 2.3. Maybe log a message
    #
    # maybe_log(msg).  Write the message to the log (using the write()
    # method), but only if its priority is higher than self.priority.

    def maybe_log(self, msg):
        assert isinstance(msg, message.message)
        # Higher priorities have lower numbers, hence the sense of
        # this test.
        if msg.priority <= self.priority:
            self.write(msg)


    # 2.4. Write a message to the log
    #
    # write(msg).  Write the message to the log.  There is no
    # implementation in the logger class (it's an abstract class).
    # Subclasses should provide the appopriate mechanism.

    def write(self, msg):
        assert isinstance(msg, message.message)
        # logger is an abstract class; no implementation of write().
        assert 0


    # 2.5. Log a message
    #
    # log(msg). Invoke maybe_log(), with error handling. This method
    # is the external interface to the logger.  You may not override
    # this method in a subclass unless you support identical error
    # handling.  Override the maybe_log method instead.

    def log(self, msg):
        assert isinstance(msg, message.message)
        try:
            self.maybe_log(msg)
        except:
            type, value = sys.exc_info()[:2]
            self.advise_failed('%s: %s' % (type, value))


    # 2.6. Handle a log failure
    #
    # advise_failed(err).  Handle the error message by calling
    # log_failed_hook().  But only once per logger.  We will only
    # advise that a log has failed once.  Otherwise the administrator
    # could be bombarded with failure messages.

    def advise_failed(self, err):
        # Is this the first time this logger has run into problems?
        if not self.has_failed:
            self.has_failed = 1
            context = self.failure_context()
            self.log_failed_hook(err, context)


    # 2.7. User hook for log failure
    #
    # By default we just re-raise the error.  This hook can be set
    # elsewhere (by the replicator, for instance, to a method that
    # sends mail to an administrator).  For example, in a method in
    # the replicator class one could write:
    #
    #    def log_failed_hook(err, context, r=self):
    #        r.mail_report(...)
    #    self.config.logger.set_log_failed_hook(log_failed_hook)
    #
    # in order to mail the failure report to the administrator.

    def log_failed_hook(self, err, context):
        raise

    def set_log_failed_hook(self, hook):
        assert type(hook) in [types.FunctionType, types.MethodType]
        self.log_failed_hook = hook


    # 2.8. Describe context of log failure
    #
    # failure_context().  Return a message object describing the
    # context of the log failure, suitable for using as an
    # introduction to a failure report.

    def failure_context(self):
        # "An attempt to write a log message to %s failed."
        return catalog.msg(1018, self)


# 3. FILE LOGGER CLASS
#
# This subclass of logger appends messages to a file stream, or to the
# standard output if no file is specified when the instance is created.
# The output buffer is flushed after each message is written so that the
# log can be recovered even if the program crashes.

class file_logger(logger):
    file = None

    def __init__(self, file = sys.stdout, priority = message.INFO,
                 max_length = 10000):
        # Test interface, not type, since sys.stdout is not a native
        # file objectin environments like PythonWin.
        assert hasattr(file, 'write') and hasattr(file, 'flush')
        logger.__init__(self, priority, max_length)
        self.file = file

    def write(self, msg):
        assert isinstance(msg, message.message)
        self.file.write(self.format_with_date(msg))
        self.file.write('\n')
        self.file.flush()

    def failure_context(self):
        if self.file is sys.stdout:
            # "An attempt to write a log message to standard output
            # failed."
            return catalog.msg(1017)
        else:
            destination = getattr(self.file, 'name', str(self.file))
            # "An attempt to write a log message to %s failed."
            return catalog.msg(1018, destination)


# 4. SYSTEM LOGGER CLASS
#
# This subclass of logger logs messages to the system log on Unix (using
# syslog).  On other operating systems, it does nothing.

class sys_logger(logger):
    def __init__(self, priority = message.INFO, max_length = 10000):
        logger.__init__(self, priority, max_length)
        if os.name == 'posix':
            import syslog
            self.syslog = syslog.syslog
            syslog.openlog('p4dti', syslog.LOG_PID, syslog.LOG_DAEMON)
            syslog.setlogmask(syslog.LOG_UPTO(priority))

    def syslog(self, priority, text):
        pass

    # 4.1. Log a message to the system log
    #
    # We override the maybe_log() method rather than the write()
    # method because we use syslog's logmask feature instead of
    # checking the priority ourselves.

    def maybe_log(self, msg):
        assert isinstance(msg, message.message)
        self.syslog(msg.priority, self.stringify_message(msg))

    def failure_context(self):
        # "An attempt to write a log message to the system log failed."
        return catalog.msg(1019)


# 5. LOGGER CLASS FOR MULTIPLE LOGS
#
# This is a meta-logger that writes the message to each of a list of
# other loggers.  This is so that the administrator can arrange for
# messages to go to several places in the integration configuration [RB
# 2000-08-10, 5.1].

class multi_logger(logger):
    loggers = []

    def __init__(self, loggers = [], priority = message.INFO,
                 max_length = 10000):
        assert isinstance(loggers, types.ListType)
        for l in loggers:
            assert(isinstance(l, logger))
        logger.__init__(self, priority, max_length)
        self.loggers = loggers

    def maybe_log(self, msg):
        assert isinstance(msg, message.message)
        for l in self.loggers:
            l.log(msg)

    def set_log_failed_hook(self, hook):
        assert type(hook) in [types.FunctionType, types.MethodType]
        # Call superclass method.
        logger.set_log_failed_hook(self, hook)
        for l in self.loggers:
            l.set_log_failed_hook(hook)


# 6. WINDOWS EVENT LOGGER CLASS

class win32_event_logger(logger):
    # Application name.
    application = None

    # Map from message priority to Windows event type.
    event_type = None

    def __init__(self, rid, priority = message.INFO,
                 max_length = 10000):
        logger.__init__(self, priority, max_length)
        self.application = "P4DTI-" + rid
        # The "Event Message File" is eventlog.dll, in the same
        # directory as the message.py module.
        emf = os.path.join(
            os.path.dirname(os.path.abspath(message.__file__)),
            "eventlog.dll")
        import win32evtlogutil
        win32evtlogutil.AddSourceToRegistry(self.application, emf)
        import win32evtlog
        self.event_type = {
            message.EMERG:   win32evtlog.EVENTLOG_ERROR_TYPE,
            message.ALERT:   win32evtlog.EVENTLOG_ERROR_TYPE,
            message.CRIT:    win32evtlog.EVENTLOG_ERROR_TYPE,
            message.ERR:     win32evtlog.EVENTLOG_ERROR_TYPE,
            message.WARNING: win32evtlog.EVENTLOG_WARNING_TYPE,
            message.NOTICE:  win32evtlog.EVENTLOG_WARNING_TYPE,
            message.INFO:    win32evtlog.EVENTLOG_INFORMATION_TYPE,
            message.DEBUG:   win32evtlog.EVENTLOG_INFORMATION_TYPE,
            }

    def write(self, msg):
        assert isinstance(msg, message.message)
        import win32evtlogutil
        win32evtlogutil.ReportEvent(self.application, 0, 0,
                                    self.event_type[msg.priority],
                                    [self.stringify_message(msg)])

    def failure_context(self):
        # "An attempt to write a log message to the NT event log
        # failed."
        return catalog.msg(1020)


# A. REFERENCES
#
# [RB 2000-08-10] "Perforce Defect Tracking Integration Administrator's
# Guide"; Richard Brooksby; Ravenbrook Limited; 2000-08-10;
# <http://www.ravenbrook.com/project/p4dti/version/2.4/manual/ag/>.
#
# [Requirements] "Perforce Defect Tracking Integration Project
# Requirements"; Gareth Rees; Ravenbrook Limited; 2000-05-24;
# <http://www.ravenbrook.com/project/p4dti/req/>.
#
#
# B. DOCUMENT HISTORY
#
# 2000-10-16 GDR Created.
#
# 2000-11-30 GDR Added some type checking.
#
# 2001-01-19 NB Added sys_logger.
#
# 2001-01-26 NB Removed extra \n.
#
# 2001-03-01 NB Fix for job000237.
#
# 2001-03-02 RB Transferred copyright to Perforce under their license.
#
# 2001-03-11 GDR Formatted as a document.  Uses the message class.
# Check digit computation moved to message.py.  Loggers pay attention to
# message priority.
#
# 2001-04-10 NB job000292: Added call to str() as our message objects
# were upsetting syslog.syslog.
#
# 2001-09-12 GDR Added logger class for the Windows event log.
#
# 2001-11-06 NDL File logger works in PythonWin.
#
# 2001-11-20 NDL Added log_failed_hook.
#
# 2001-11-26 GDR Truncate log messages to a specified length.
#
# 2001-12-07 GDR Test files by interface, not type; don't assume it has
# a 'name' attribute.  Allow log failure hooks to be methods.
#
#
# C. COPYRIGHT AND LICENSE
#
# This file is copyright (c) 2001 Perforce Software, Inc.  All rights
# reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
# 1.  Redistributions of source code must retain the above copyright
#     notice, this list of conditions and the following disclaimer.
#
# 2.  Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in
#     the documentation and/or other materials provided with the
#     distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# HOLDERS AND CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
# OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
# TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
# USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
# DAMAGE.
#
#
# $Id: //info.ravenbrook.com/project/p4dti/version/2.4/code/replicator/logger.py#2 $
